### ISSUE TITLE
undefined from useServerSideMutation()  /  DevalueError: Cannot stringify arbitrary non-POJOs

### ISSUE THREAD

**Issue by rondonjon** [rakkasjs/rakkasjs#113]
In an attempt to create a cookie / session middleware for the page and the API routes in my application, I have created a `src/routes/hattip-entry.ts` file as follows:

```
import { cookie } from "@hattip/cookie"
import { createRequestHandler, RequestContext } from "rakkasjs"
import { ID } from "src/crud/id"
import { getSessionById, Session } from "src/crud/session"

declare module "rakkasjs" {
  interface ServerSideLocals {
    session?: Session
    setSession: (session: Session | undefined) => void
  }
}

const COOKIE_SESSION_ID = "session_id"

export default createRequestHandler({
  middleware: {
    beforePages: [
      cookie(),
      async (ctx: RequestContext) => {
        const sessionId = ctx.cookie[COOKIE_SESSION_ID]
        let session: Session | undefined

        if (typeof sessionId === "string") {
          session = getSessionById(sessionId as ID, true)
        }

        ctx.locals.session = session

        ctx.locals.setSession = (session: Session | undefined) => {
          if (session) {
            ctx.setCookie(COOKIE_SESSION_ID, session.id, {
              httpOnly: true,
              maxAge: 60 * 60,
              sameSite: "strict",
              secure: true,
            })
          } else {
            ctx.deleteCookie(COOKIE_SESSION_ID)
          }
        }

        return ctx.next()
      },
    ],
  },
})
```

I have a server-side mutation in which I am trying to update these session fields:

```
export const useLogin = () => {
  const queryClient = useQueryClient()

  const m = useServerSideMutation<MutationResult, MutationParams>(
    async (context, { email, password }) => {
      if (context.locals.session) {
        terminateSession(context.locals.session.id)
      }

      const user = getUserFacetByAuth(email, password)

      if (user) {
        const session = createSession(user.id)
        context.locals.setSession?.(session)
        return {
          session,
        }
      } else {
        context.locals.setSession?.(undefined)
        return {
          error: new Error("failed"),
        }
      }
    },
    {
      onSettled: () => {
        queryClient.invalidateQueries("authUser")
      },
    }
  )

  if (!m) {
    return {
      error: undefined,
      isFetching: false,
      login: () => undefined,
    }
  }

  const { mutate, isLoading, data, error } = m

  return {
    error: error || data,
    isFetching: isLoading,
    login: mutate,
  }
}
```

There are a couple of weird things happening now.

The first is that I am getting `undefined` as return value from `useServerSideMutation()` when the code is running on the server side. According to the `.d.ts` though, this should never happen, and the result should always be an object with `mutate()` etc. on it.

To work around this problem, I have temporarily patched the code to check for `!m` above and return a dummy result on `undefined`.

But it's still not working. I am now getting a `DevalueError` in the server-side console, and this one leaves me clueless:

```
DevalueError: Cannot stringify arbitrary non-POJOs
    at walk (file:///.../node_modules/devalue/src/uneval.js:86:13)
    at walk (file:///.../node_modules/devalue/src/uneval.js:101:7)
    at Proxy.uneval (file:///.../node_modules/devalue/src/uneval.js:108:2)
```

Any idea what could be causing these problems?

**Comment by cyco130** [comment]
The bug you describe (`useMutation` returning `undefined` on the server) is solved by #104. It is in the `@next` version (i.e. npm version `0.7.0-next.10`) but I just merged to the stable branch too. It should be solved if you upgrade to 0.6.20. But I recommend you start using `@next` at this point. 

The other problem happens because `useServerSide(Query|Mutation)` uses [devalue](https://github.com/Rich-Harris/devalue) to serialize their return value to be sent from the server to the client and `devalue` only supports primitives and plain data objects. `ctx.session` is not a plain data object (it has methods, but `ctx.session.data` *is* OK to serialize). See the relevant paragraph [from the docs](https://rakkasjs.org/guide/use-server-side-query):

> Rakkas uses [`@brillout/json-serializer`](https://github.com/brillout/json-serializer) to serialize the captured values and [devalue](https://github.com/Rich-Harris/devalue) to serialize the return value. They both support `Date`, `undefined`, `Set`, `Map`, `BigInt`, `RegExp`, `NaN`, and `Infinity` in addition to standard JSON-serializable types. `devalue` also supports repeated and cyclic references. Captured values and the return value must be serializable with the respective libraries. In particular, functions, arbitrary class instances, DOM elements, etc. cannot be serialized properly.

**Comment by rondonjon** [comment]
> The bug you describe (useMutation returning undefined on the server) is solved by https://github.com/rakkasjs/rakkasjs/pull/104.

Thanks @cyco130. Upgrading solved the problem. Works now!
